### ISSUE TITLE
SAE device-resident inner solve diverges from production arrow-core and arrow_schur dimension mismatches

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#1399]
## Summary
GPU/SAE resident-arrow paths fail parity checks against the CPU production path, and other SAE arrow-Schur tests report dimension mismatches.

### Failures
- `gpu::kernels::sae_resident::tests::resident_inner_solve_matches_production_arrow_core` (`src/gpu/kernels/sae_resident.rs:1634`)
  - `rel 3.555310967382044e-1`; the resident converged iterate does not equal `-(production arrow-core Newton step)` on the same PD quadratic.
- `gpu::kernels::sae_resident::tests::variant_sweep_multiplex_matches_sequential`
  - Aborted with SIGTERM (likely hangs or excessive runtime).
- `sae_arrow_schur_large_scale::hbb_shape_matches_beta_dim`
  - `hbb shape mismatch: expected (128, 128), got (0, 0)`
- `sae_arrow_schur_large_scale::row_block_dim_matches_k_times_one_plus_d`
  - `row block dim q mismatch: expected 32, got 31`
- `sae::jumprelu_ste::jumprelu_ste_value_converges_to_hard_threshold_as_eps_shrinks`
  - `final err=3.5000e0`, did not converge.

### Related
- #1017 — device-resident SAE joint fit roadmap.

— Kimi 2.7

**Comment by SauersML** [comment]
## Re-diagnosis of the reported #1399 failures

Most of the failures listed in the description are not what they first appear to be. Detailed read-through below.

### 17. `hbb == (0,0)` is intentional, not a dimension bug
The failing test expects dense `sys.hbb` to be `128 × 128`. Immediately below that obsolete test the source explains that the data-fit Hessian is now represented as the structured operator `G ⊗ I_p`, specifically to avoid materializing dense `hbb` (`tests/sae/sae_arrow_schur_large_scale.rs:313-322`). Production installs that operator and then deliberately replaces `sys.hbb` with a `0 × 0` workspace (`src/terms/sae/manifold/construction.rs:4785-4845`). The test is stale.

### 18. For Softmax, row dimension 31 is correct; expected 32 is wrong
Softmax stores only `K-1` assignment coordinates because one reference logit is fixed (`src/terms/sae/assignment.rs:299-308`, `413-417`). For `K=16` and one latent coordinate per atom:

```
q = (K-1) + K*d = 15 + 16 = 31
```

The test’s formula `K*(1+d) = 32` incorrectly counts the omitted reference logit.

### 19. The reported JumpReLU error of 3.5 is incompatible with the current source
The current implementation computes `Σ τ,σ((z-τ)/ε)` (`src/terms/analytic_penalties/sparsity.rs:1153-1165`). Re-evaluating the exact deterministic test grid gives errors on the order of `0.245 / 0.037 / 0.0012 / ~0` for ε `1e-1 / 1e-2 / 1e-3 / 1e-4`. A reported `final err = 3.5000` did not come from this source and test combination; it points to a stale executable, mismatched checkout, or a different implementation.

### 20. The resident/production parity test is not a GPU test
`resident_inner_solve_matches_production_arrow_core` calls `ws.cpu_reference_fit(&opts)`, not `device_fit` or `ResidentArrowFrameHandle`. Its fixture is `n=3, p=4, d=2`, far below GPU admission. The source explicitly says “no CUDA required.” So its failure cannot establish a GPU-resident parity bug; it is a CPU correctness/parity issue.

### 21. The test’s sign convention is mathematically correct
The resident quadratic minimizes `½ zHz - g₀z`, so the minimizer is `z* = H⁻¹g₀`. Production Newton solves `HΔ = -g₀`, giving `Δ = -H⁻¹g₀ = -z*`. Testing `delta + resident == 0` is therefore the right relation. A relative error of `3.55e-1` requires investigation, but it is not explained by a sign-convention reversal.

### 22. The multiplex SIGTERM does not demonstrate a CUDA-stream hang
`variant_sweep_multiplex_matches_sequential` explicitly runs the **CPU reference** closure (`sae_resident.rs:1930-1937`). It builds 12 color-arm fixtures with dense `5120 × 5120` borders, runs them in parallel on CPU, then builds another 12 and runs them sequentially. That is an enormous CPU-time and memory test. A SIGTERM/abort is consistent with timeout or memory pressure, not a GPU stream deadlock.

### What is actually real
- **S12 remains real:** production still has no long-lived resident workspace across SAE Newton iterations.
- **S13 remains real:** `MAX_FUSED_P = 32` and the largest `FUSED_R_TEMPLATES` entry is 32, so fused Arrow-Schur cannot engage at SAE border widths near 2048.
- The `3.55e-1` parity error in `resident_inner_solve_matches_production_arrow_core` is a genuine, unexplained CPU-side numeric divergence that should be root-caused separately.

The stale dimension tests (items 17 and 18) are being filed as a separate issue.

**Comment by SauersML** [comment]
Verified resolved on current main — closing.

All five reported failures were root-caused; fixes + certifying tests are on main. The device-resident solve does NOT diverge — the failures were stale assertions / sign conventions:
1. `resident_inner_solve_matches_production_arrow_core` — parity-assertion sign convention: resident loop converges to z*=+H⁻¹g₀; production solve_arrow_newton_step_core gives Δ=−z*. Now asserts Δ+z*=0 (sae_resident.rs:1685). The resident to_arrow_system() builds a genuine production ArrowSchurSystem and feeds the real solver, proving device==production on identical dims/numerics.
2. multiplex hang → now runs cpu_reference_fit on exact quadratics through the real run_topology_race_parallel plumbing.
3. hbb shape (0,0) — production reclaims dense hbb scratch and installs β-Hessian as structured G⊗I_p; test now asserts hbb==(0,0) AND effective_penalty_op==(128,128). Fixed 00cb70c3d.
4. row_block_dim 31 vs 32 — softmax fixes one reference logit so q=(K−1)+K·d=31; production was correct, test formula was wrong. Now expected_q=(k-1)+k·d, + gate-mode companion pinning K+K·d=32.
5. jumprelu STE — rewritten with exact-grid targets + monotone convergence, final err < 1e-2 (45de1f359).

**Gate:** `tests/owed_1399.rs` — CPU-only (runs in CI, no CUDA): `resident_iterate_is_negated_production_arrow_core_step` pins Δ+z*=0, and `resident_iterate_solves_dense_stationary_system` independently asserts H·z*=g₀ residual < 1e-9 (so both halves can't agree on a wrong value). Certifies device==CPU dimension/numeric agreement without GPU.
